trieste/space.py

```diff
@@ -329,3 +346,124 @@ def __mul__(self, other: Box) -> Box:
 
     def __deepcopy__(self, memo: dict[int, object]) -> Box:
         return self
+
+
+class ProductSearchSpace(SearchSpace):
+    r"""
+    Product :class:`SearchSpace` consisting of a product of
+    multiple :class:`SearchSpace`s. This class provides functionality for
+    accessing either the resulting combined search space or each individual space.
+
+    Note that this class assumes that individual points in product spaces are
+    represented with their inputs in the same order as specified when initializing
+    the space.
+    """
+
+    def __init__(self, spaces: list[tuple[str, SearchSpace]]):
+        r"""
+        :param spaces: A list of tag-space tuples summarizing the names and domains of
+            the space's subspaces.
+        """
+
+        tags = [tag for (tag, space) in spaces]
```

Comment:
Better store these in a tuple, so nobody can modify it by mistake (since you're exposing it in a property):

```suggestion
        tags = tuple(tag for (tag, space) in spaces)
```

Reply:
nice trick!